### ISSUE TITLE
Interface for optional unfree dependencies

### ISSUE THREAD

**Issue by davidak** [nixpkgs-architecture/issues#18]
**Issue description**

There are packages that have optional unfree dependencies.

As far as i know, there is no recommended way to handle that.

This issue came up in https://github.com/NixOS/nixpkgs/pull/221861/files#r1143804170

cc @RaitoBezarius @alyssais

**Goal**

I think we need a documented way to handle this, so packages behave consistently.

I think good UX would be that optional unfree dependencies are used when my system allows it by `nixpkgs.config.allowUnfree` or for specific dependencies with `allowUnfreePredicate` or `allowlistedLicenses` etc.

Maybe you can move this logic to a nixpkgs library function, so the code in a package is simple.

It seem also useful to be able to allow it per package with an option like `enableUnfree`.

It might also be useful to have a package flavor installable as `packageUnfree`.

It should be simple and obvious how a user can install a package with optional unfree dependencies.

**Comment by alyssais** [comment]
> I think we need a documented way to handle this, so packages behave consistently.

As far as I can tell, we are already handling this consistently.  There are no packages in Nixpkgs that I can find that modify their behaviour based on `config.allowUnfree`, and there's a good reason for that: if they did, people with `allowUnfree` enabled wouldn't benefit from Hydra builds.  I really don't think there's anything to do here, other than to document the reality that we don't modify packages based on `config.allowUnfree`.

**Comment by davidak** [comment]
let's talk about UX first and then about implementation detail

my suggestion is that there should be an option to change the default of `allowUnfree` for **all packages** (that have that option) at once, similar to how `hardware.enableAllFirmware` does it for firmware. `nixpkgs.config.allowUnfree` might have a different purpose (just allowing it, but not actually enabling), so we need a new option for that

With that people that have no problem with unfree software can enable this one checkbox in the nixos installer and get all the unfree stuff we have to offer. There could also be a second checkbox to "Use optional unfree dependencies" with a warning about the local rebuilds and a link to instructions how to enable it per package otherwise.

![Screenshot from 2023-03-21 23-28-47](https://user-images.githubusercontent.com/91113/226756508-48c84cfe-83da-4a2f-9527-d849b9a73e9b.png)

That would be aligned with a goal to empower the user to use their computer as they want, but conflicts with a goal to push free software (by making it hard to use unfree stuff).

**Comment by RaitoBezarius** [comment]
With the current nixpkgs, this seems quite hard to achieve without some sort of overlay which would replace each free version with the closest unfree version of a software.
With a future nixpkgs relying on a module system, this is somewhat trivial.

In all the cases, before discussing "unfree" dependencies, there must be probably a discussion on how a end-user discovers optional dependencies for a tool beyond reading the source code and how to present the multiple options.

Then, once we have a solution taking into account this problem, we can solve the optional *unfree* dependency problem by applying this policy thing.

Though, it will introduce an extra headache and something that I suppose will be annoying for end-users: there is no cache for those packages, so it means rebuilding most of the time each time the dependency has to be rebuilt. I personally want to avoid this.

So any policy should reflect the tension between local rebuilds and "optional unfree" or unfree packages.

**Comment by davidak** [comment]
>With the current nixpkgs, this seems quite hard to achieve

No, you only have to replace `enableUnfree ? false` with `enableUnfree ? config.nixpkgs.config.enableUnfree` (example name) in the 8 packages that have it. Those with true can keep that. Trivial change.

>before discussing "unfree" dependencies, there must be probably a discussion on how a end-user discovers optional dependencies

that's an unrelated topic which does not block this one. the current situation is they don't

>it will introduce an extra headache

yes, and that should be made very clear. and users will still complain and think nixos is bad

the current situation is, you have to enable it for every package separately. which is also a headache i think
but maybe even if a user don't care about licenses, the free version of a package is good enough in 99% of situations and enabling it manually is totally fine in the 1% situations where specific features are needed.

**Comment by RaitoBezarius** [comment]
> > With the current nixpkgs, this seems quite hard to achieve
> 
> No, you only have to replace `enableUnfree ? false` with `enableUnfree ? config.nixpkgs.config.enableUnfree` (example name) in the 8 packages that have it. Those with true can keep that. Trivial change.

My whole point is to say that this is a problem to do this.
Unfree taints a package and I would argue it is better to separate them and consider them as separate packages under different names.

It is okay to disagree on this concept of having multiple identities for package, but I am not convinced that making package silently change of nature completely based on a meta configuration flag of nixpkgs is a good idea.

I would rather consider {options of packages} ('minimal', 'normal', 'bloated') → {unfree, free} → package (whatever the order of the arguments).

> > before discussing "unfree" dependencies, there must be probably a discussion on how a end-user discovers optional dependencies
> 
> that's an unrelated topic which does not block this one. the current situation is they don't

it's an important topic because you argue in the bottom that "if a user don't care about licenses, blablabla"
but the whole question is : to what audience do we cater regarding this feature?

I mean, if enabling manually in the 1% situation is fine, then, I suppose that it's not a big deal to do `(xxx.override { enableUnfree = true; })` or `xxxUnfree` rather than `xxx`, questioning what do we even want to do here.

> > it will introduce an extra headache
> 
> yes, and that should be made very clear. and users will still complain and think nixos is bad

I mean, we have the potential to try to find the least painful path and the one which communicate properly the intents and effects to end users. So I'm not sure we need to proceed with something we think will pose problems.

> the current situation is, you have to enable it for every package separately. which is also a headache i think but maybe even if a user don't care about licenses, the free version of a package is good enough in 99% of situations and enabling it manually is totally fine in the 1% situations where specific features are needed.

I don't agree with the assessment of the situation. I run `ripgrep` on nixpkgs and I think the situation is much more nuanced.

You have plenty of different situations: drone does `enableUnfree = true;` by default and there is a drone-oss package. Same for elasticsearch versions, same for logstash, Kibana has no OSS version, it's unfree or nothing. OpenCV disable unfree by default. 7z too. libav requires unfree if you enable FAAC support. VirtualBox has unfree disabled by default.
For the SAT solver glucose, you have two identities: glucose (OSS) and glucose-syrup (unfree), correctly reflected.

I don't see any package where enabling "separately" is a pain, most of them requires an active consideration from the user, e.g. OpenCV, VirtualBox, libav, glucose.

Certainly, given the few of them, writing once the invocation to enable the unfreeness is not a big deal?
You say that "free version of a package is good enough", I'm not sure, at least, this is not consistent in nixpkgs as we have `-oss` versions rather than `-unfree` version. :)

**Comment by Aleksanaa** [comment]
A point that may not be accurate: if there is an global option to turn on "enableUnfree" for all packages (if the option exists), the name of the option should be "preferUnfree", not "enableUnfree" or "allowUnfree". As NixOS is a community-driven distribution, not a commercial one, it's probably not one of the development goals to prefer opting in proprietary source. On the contrary, "enableUnfree" just provides a little convenience for the user, making it easier to fill in the lack of functionality (and this lack may not affect the use of other users).

By the way, I don't think anyone will actually say "hey, if there are proprietary sources, please try to bring them, the more complete the better", right?
